physx-sys/build.rs

```diff
@@ -1,166 +1,83 @@
-        // Get all sub-directories in "bin"
-        let out_dirs = entries
-            .filter_map(std::result::Result::ok)
-            .filter(|entry| entry.metadata().map(|m| m.is_dir()).unwrap_or(false));
-
-        // Find the ones containing "debug" or "profile", and pick the most recently modified.
-        let out_dir = out_dirs
-            .filter_map(|out_dir| {
-                let mut out_dir = out_dir.path();
-                out_dir.push(build_profile);
-
-                let newest_in_dir = most_recent_modification_time_in_dir(&out_dir)?;
-                Some((out_dir, newest_in_dir))
-            })
-            .max_by_key(|(_, m)| *m)
-            .map(|(d, _)| d);
-
-        out_dir
-            .expect("could not locate output directory for PhysX static libs")
-            .display()
-            .to_string()
-    } else {
-        panic!("Could not inspect cmake build output directory");
-    }
-}
+include!("cc.rs");
+include!("cmake.rs");
 
 fn main() {
-    let physx_root_dir = env::current_dir().unwrap().join("PhysX/physx");
-    let physx_parent_dir = physx_root_dir.parent().unwrap();
-
-    // Based on PhysX/physx/generate_projects.sh
-    env::set_var(
-        "PHYSX_ROOT_DIR",
-        &physx_root_dir.to_string_lossy().into_owned(),
-    );
-    env::set_var(
-        "PM_PxShared_PATH",
-        physx_parent_dir
-            .join("pxshared")
-            .to_string_lossy()
-            .into_owned(),
-    );
-    env::set_var(
-        "PM_CMakeModules_PATH",
-        physx_parent_dir
-            .join("externals/cmakemodules")
-            .to_string_lossy()
-            .into_owned(),
-    );
-    env::set_var(
-        "PM_opengllinux_PATH",
-        physx_parent_dir
-            .join("externals/opengl-linux")
-            .to_string_lossy()
-            .into_owned(),
-    );
-    env::set_var("PM_PATHS", env::var("PM_opengllinux_PATH").unwrap());
-
-    let output_dir = env::var("OUT_DIR").unwrap();
-    let output_dir_path: PathBuf = output_dir.clone().into();
-
-    let build_mode = if cfg!(debug_assertions) {
-        "debug"
-    } else {
-        "profile"
+    // Use the optimization level to determine the build profile to pass, we
+    // don't use cfg!(debug_assertions) here because I'm not sure what happens
+    // with that when build dependencies are configured to be debug and the
+    // actual target is meant to be release, so this seems safer
+    let build_mode = match env::var("OPT_LEVEL")
+        .ok()
+        .and_then(|s| s.parse::<i32>().ok())
+        .unwrap_or(1)
+    {
+        0 => "debug",
+        _ => "profile",
     };
 
-    let target_os = get_physx_target_os_name();
-    let target_os = target_os.as_str();
-    let mut physx_cfg = Config::new("PhysX/physx/source/compiler/cmake");
-
-    if target_os == "linux" {
-        physx_cfg.define("CMAKE_LIBRARY_ARCHITECTURE", "x86_64-unknown-linux-gnu");
-    }
+    let use_cmake = env::var("CARGO_FEATURE_USE_CMAKE").is_ok();
```

Comment:
This is potentially a bit confusing -- other cargo/rust flags use the `=0` and `=1` convention to disable and enable a feature. Here you would need to actually unset the var.

Reply:
This is just due to how features appear in the environment.

> `CARGO_FEATURE_<name>` — For each activated feature of the package being built, this environment variable will be present where <name> is the name of the feature uppercased and having - translated to _.

Reply:
Ah, I see. So one would never set this by hand. Cool, sorry for the confusion then.

Reply:
No worries, glad you're keeping me honest!